src/server/pps/server/api_server.go

```diff
@@ -464,6 +467,248 @@ func (a *apiServer) RestartDatum(ctx context.Context, request *pps.RestartDatumR
+			},
+		},
+	)
+	if err != nil {
+		return nil, err
+	}
+	oldFile := &pfs.File{
+		Commit: commitInfo.ParentCommit,
+		Path:   fmt.Sprintf("/%v", request.Job.ID),
+	}
+	resp, err := pfsClient.DiffFile(ctx, &pfs.DiffFileRequest{newFile, oldFile, true})
+	if err != nil {
+		return nil, err
+	}
+
+	// The newFileInfos contain datums that were new in this job
+	// For these datums, populate the status and stats
+	blacklist := map[string]bool{
+		"stats": true,
+		"logs":  true,
+		"pfs":   true,
+	}
+	pathToDatumHash := func(path string) (string, error) {
+		_, datumHash := filepath.Split(path)
+		if _, ok := blacklist[datumHash]; ok {
+			return "", fmt.Errorf("value %v is not a datum hash", datumHash)
+		}
+		return datumHash, nil
+	}
+	var egGetDatums errgroup.Group
+	getDatumsSemaphore := make(chan struct{}, 200)
+	var datumsMutex sync.Mutex
+	for _, fileInfo := range resp.NewFiles {
+		egGetDatums.Go(func() error {
+			getDatumsSemaphore <- struct{}{}
+			defer func() { <-getDatumsSemaphore }()
+			fileInfo := fileInfo
+			datumHash, err := pathToDatumHash(fileInfo.File.Path)
+			if err != nil {
+				// not a datum, nothing to do here
+				return nil
+			}
+			datum, err := a.getDatum(ctx, jobInfo.OutputCommit.Repo.Name, statsBranch.Head, request.Job.ID, datumHash)
+			datumsMutex.Lock()
+			defer datumsMutex.Unlock()
+			datums[datumHash] = datum
+			return err
+		})
+	}
+	err = egGetDatums.Wait()
+	if err != nil {
+		return nil, err
+	}
+
+	var datumInfos []*pps.DatumInfo
+	for datumHash, datum := range datums {
+		if _, ok := blacklist[datumHash]; ok {
+			// not a datum
+			continue
+		}
+		datumInfos = append(datumInfos, datum)
+	}
+
+	// Sort results (failed first, slow first)
+	sort.Sort(byDatumStateThenTime(datumInfos))
+
+	return &pps.DatumInfos{datumInfos}, nil
+}
+
+type byDatumStateThenTime []*pps.DatumInfo
+
+func (a byDatumStateThenTime) Len() int      { return len(a) }
+func (a byDatumStateThenTime) Swap(i, j int) { a[i], a[j] = a[j], a[i] }
+func (a byDatumStateThenTime) Less(i, j int) bool {
+	byState := a[i].State < a[j].State
+	if a[i].State != a[j].State {
+		return byState
+	}
+	if a[i].Stats == nil || a[j].Stats == nil {
+		return byState
+	}
+	return client.GetDatumTotalTime(a[i].Stats) > client.GetDatumTotalTime(a[j].Stats)
+}
+
+func (a *apiServer) getDatum(ctx context.Context, repo string, commit *pfs.Commit, jobID string, datumID string) (*pps.DatumInfo, error) {
+	pfsClient, err := a.getPFSClient()
+	if err != nil {
+		return nil, err
+	}
+	// Populate status
+	state := pps.DatumState_FAILED
+	stateFile := &pfs.File{
+		Commit: commit,
+		Path:   fmt.Sprintf("/%v/%v/failure", jobID, datumID),
+	}
+	getFileClient, err := pfsClient.GetFile(ctx, &pfs.GetFileRequest{stateFile, 0, 0})
+	if err := grpcutil.WriteFromStreamingBytesClient(getFileClient, ioutil.Discard); err != nil {
+		if strings.Contains(err.Error(), "not found") {
+			state = pps.DatumState_SUCCESS
+		} else {
+			return nil, err
+		}
+	}
```

Comment:
Are you just trying to check for the existence of the `failure` file here? If so `InspectFile` is probably an easier way to do that.